File: launcher/driver-api/jvm/src/main/java/dev/argon/driver/api/CompilerDriverOptions.java
package dev.argon.driver.api;

import java.io.*;
import java.nio.file.Path;
import java.util.List;

import dev.argon.backend.api.*;
import dev.argon.driver.api.command.DriverCommand;

public final class CompilerDriverOptions {
    public CompilerDriverOptions(
		List<BackendFactory> backendFactories,
		DriverCommand<
			BinaryResource<IOException>,
			DirectoryResource<IOException>,
			BinaryResourceSink<IOException>,
			DirectoryResourceSink<IOException>
		> command,
		InputStream stdin,
		PrintStream stdout,
		PrintStream stderr
    ) {
        this.backendFactories = backendFactories;
        this.command = command;
		
	    this.stdin = stdin;
	    this.stdout = stdout;
	    this.stderr = stderr;
    }
    
    private final List<BackendFactory> backendFactories;
    private final DriverCommand<
	    BinaryResource<IOException>,
	    DirectoryResource<IOException>,
	    BinaryResourceSink<IOException>,
	    DirectoryResourceSink<IOException>
    > command;
	private final InputStream stdin;
	private final PrintStream stdout;
	private final PrintStream stderr;

	public List<BackendFactory> getBackendFactories() {
		return backendFactories;
	}
	
    public DriverCommand<
	    BinaryResource<IOException>,
	    DirectoryResource<IOException>,
	    BinaryResourceSink<IOException>,
	    DirectoryResourceSink<IOException>
    > getCommand() {
        return command;
    }
	
	public InputStream getStdin() {
		return stdin;
	}
	
	public PrintStream getStdout() {
		return stdout;
	}
	
	public PrintStream getStderr() {
		return stderr;
	}
    
}

File: launcher/driver/node/src/main/scala/dev/argon/driver/sjs/CompilerDriverOptions.scala
package dev.argon.driver.sjs

import dev.argon.backend.sjs.{BackendFactory, BinaryResource, DirectoryResource, BinaryResourceSink, DirectoryResourceSink}
import dev.argon.driver.sjs.command.DriverCommand
import dev.argon.io.PathLike

import scala.scalajs.js

trait CompilerDriverOptions extends js.Object {
  def backendFactories: js.Array[BackendFactory]
  def command: DriverCommand[BinaryResource[js.Error], DirectoryResource[js.Error], BinaryResourceSink[js.Error], DirectoryResourceSink[js.Error]]
}
